### PR TITLE
feat: add scorecards analysis workflow

### LINKED ISSUE #32: Add Scorecard security checks
It would be nice to add [Scorecard analysis](https://github.com/ossf/scorecard) to this repo as a GitHub Actions workflow to compute the security score. Scorecard runs various simple checks, e.g., whether `Branch-Protection` and `Security-Policy` exist in the repo, and computes a score of 0-1. 

Here is the result that I get by locally running Scorecard:

```
RESULTS
-------
Finished [Binary-Artifacts]
Finished [Maintained]
Finished [Packaging]
Finished [CI-Tests]
Finished [Pinned-Dependencies]
Finished [Signed-Releases]
Finished [Vulnerabilities]
Finished [Branch-Protection]
Finished [Contributors]
Finished [Fuzzing]
Finished [Token-Permissions]
Finished [Security-Policy]
Finished [CII-Best-Practices]
Finished [Code-Review]
Finished [Dependency-Update-Tool]
Finished [SAST]
Aggregate score: 4.5 / 10

Check scores:
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
|  SCORE  |          NAME          |             REASON             |                                               DOCUMENTATION/REMEDIATION                                               |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts       | no binaries found in the repo  | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#binary-artifacts       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 1 / 10  | Branch-Protection      | branch protection is not       | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#branch-protection      |
|         |                        | maximal on development and all |                                                                                                                       |
|         |                        | release branches               |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | CI-Tests               | 14 out of 14 merged PRs        | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#ci-tests               |
|         |                        | checked by a CI test -- score  |                                                                                                                       |
|         |                        | normalized to 10               |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | CII-Best-Practices     | no badge found                 | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#cii-best-practices     |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 9 / 10  | Code-Review            | GitHub code reviews found for  | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#code-review            |
|         |                        | 13 commits out of the last 14  |                                                                                                                       |
|         |                        | -- score normalized to 9       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Contributors           | 0 different companies found -- | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#contributors           |
|         |                        | score normalized to 0          |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Dependency-Update-Tool | update tool detected           | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#dependency-update-tool |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Fuzzing                | project is not fuzzed in       | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#fuzzing                |
|         |                        | OSS-Fuzz                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 6 / 10  | Maintained             | 8 commit(s) found in the last  | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#maintained             |
|         |                        | 90 days -- score normalized to |                                                                                                                       |
|         |                        | 6                              |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| ?       | Packaging              | no published package detected  | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#packaging              |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 5 / 10  | Pinned-Dependencies    | dependency not pinned by hash  | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#pinned-dependencies    |
|         |                        | detected -- score normalized   |                                                                                                                       |
|         |                        | to 5                           |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | SAST                   | SAST tool is not run on all    | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#sast                   |
|         |                        | commits -- score normalized to |                                                                                                                       |
|         |                        | 0                              |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Security-Policy        | security policy file not       | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#security-policy        |
|         |                        | detected                       |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Signed-Releases        | 0 out of 2 artifacts are       | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#signed-releases        |
|         |                        | signed -- score normalized to  |                                                                                                                       |
|         |                        | 0                              |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 0 / 10  | Token-Permissions      | non read-only tokens detected  | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#token-permissions      |
|         |                        | in GitHub workflows            |                                                                                                                       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| 10 / 10 | Vulnerabilities        | no vulnerabilities detected    | https://github.com/ossf/scorecard/blob/54f1429eaa39592e61201497b108176aa2e545cc/docs/checks.md#vulnerabilities        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------------------------------------------|
```

### PR DESCRIPTION
**Pull request by behnazh**
This PR adds [Scorecards security Actions workflow](https://github.com/ossf/scorecard-action), which analyzes the repo for various supply-chain security issues.

Closes #32

### CODE REVIEW

**Comment by behnazh** [comment]
@jenstroeger note [the existing alerts](https://github.com/jenstroeger/python-package-template/pull/105/checks?check_run_id=4859270691), which probably need to be addressed before moving on :wink:

**Comment by jenstroeger** [comment]
Oh this is exciting, and interesting issues that Scorecard finds!

> […] which probably need to be addressed before moving on 😉

Hehehe very much yes.

The YAML configuration file isn’t reformatted though, is it?

**Review by jenstroeger — COMMENTED** [review]
Oh this is exciting, and interesting issues that Scorecard finds!

> […] which probably need to be addressed before moving on 😉

Hehehe very much yes.

The YAML configuration file isn’t reformatted though, is it?

**Comment by behnazh** [comment]
> The YAML configuration file isn’t reformatted though, is it? (See also PR #91.)

Ah, I thought the linting pre-commit hooks had already been merged, but I was wrong. Well, I will wait for #91 to be merged and then lint based on the latest hooks.

**Comment by jenstroeger** [comment]
@behnazh perhaps we should merge PR https://github.com/jenstroeger/python-package-template/pull/133 into this branch?

**Comment by behnazh** [comment]
TODO: change the general permissions for GitHub Actions to `Read repository contents permission` in the repo. Once this PR is ready, the workflows should not need `write` permissions by default.

**Review by jenstroeger — COMMENTED** [review]
Looking forward to the final PR 🤓

**Comment by behnazh** [comment]
The Scorecard badge is not available yet. We will add it later when they publish it.

File: .github/workflows/check-and-release-main.yaml [hunk 1]
```diff
@@ -33,15 +41,22 @@ jobs:
     - name: Build the docs
       run: make docs
   release:
+    permissions:
+      contents: write  # for relekang/python-semantic-release to push release notes and new commits/tags
```
Comment (jenstroeger) [inline .github/workflows/check-and-release-main.yaml:45]: I wonder if this one will also trigger the publish event?

File: .github/workflows/scorecards-analysis.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        results_file: results.sarif
+        results_format: sarif
+          # Read-only PAT token. To create it,
+          # follow the steps in https://github.com/ossf/scorecard-action#pat-token-creation.
+        repo_token: ${{ secrets.SCORECARD_READ_TOKEN }}
+          # Publish the results to enable scorecard badges. For more details, see
+          # https://github.com/ossf/scorecard-action#publishing-results.
+          # For private repositories, `publish_results` will automatically be set to `false`,
+          # regardless of the value entered here.
+        publish_results: true
```
Comment (jenstroeger) [inline .github/workflows/scorecards-analysis.yaml:46]: OCD: Unindent the comments by two lines, so they match the YAML entries they relate to?

**Review by jenstroeger — APPROVED** [review]
Looks good to me, although the Windows checks [still fail](https://github.com/jenstroeger/python-package-template/runs/6459769128?check_suite_focus=true).

File: .github/workflows/check-and-release-main.yaml (hunk 1 above)
Comment (behnazh) [reply .github/workflows/check-and-release-main.yaml:45]: I have tested this permission and it is enough to publish the package.

Comment (jenstroeger) [reply .github/workflows/check-and-release-main.yaml:45]: We’ll find out!

**Review by jenstroeger — APPROVED** [review]

**Comment by jenstroeger** [comment]
@behnazh, I merged PR https://github.com/jenstroeger/python-package-template/pull/194 into `staging`. Let’s hope that after you rebase this PR the Action on Windows runners works again… 🤞🏼

**Comment by behnazh** [comment]
> @behnazh, I merged PR #194 into `staging`. Let’s hope that after you rebase this PR the Action on Windows runners works again… 🤞🏼

Looks like we are good to go!

**Comment by jenstroeger** [comment]
Hooray! Merge?

**Comment by behnazh** [comment]
> Hooray! Merge?



> Hooray! Merge?

Sure :+1:

**Pull request merged**